### PR TITLE
GEODE-6150_2: SSL selector impacts

### PR DESCRIPTION
**Pull request by mivanac**
Added modification to support selector mode for SSL connections.
Added impacts to use NioFilter in selector mode, for SSL and non-SSL connections.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [*] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by onichols-pivotal** [comment]
I'm not really clear what I'm looking at here.  Given zero description of the intended change beyond "SSL selector impacts" it's hard to tell, but many of the changes look more like a merge-gone-bad...

**Comment by mivanac** [comment]
Main difference to previous PR, is modification of Message class, to use generic NioFilter, not just NioSslEngine.
I will add description.

**Comment by mivanac** [comment]
Also you should read RFC https://cwiki.apache.org/confluence/display/GEODE/Server-Side+Thread+Pooling+for+Client-Server+TLS before reviewing this PR.

**Comment by DonalEvans** [comment]
Looking at the changes included in this PR, it seems like something has gone wrong with rebasing/merging the git branch. It looks like at least 4 of the most recent commits to develop are somehow also included in this change set.

**Review by echobravopapa — CHANGES_REQUESTED** [review]
@mivanac what's the plan for getting the RFC discussed? https://cwiki.apache.org/confluence/display/GEODE/Server-Side+Thread+Pooling+for+Client-Server+TLS

the RFC at present is a placeholder, awaiting your expansion of the your plan on all of this...

**Comment by mivanac** [comment]
There was no comments on this RFC until now, only comments on previous PRs. So I don`t know what are we discussing, potential new alternatives? Current PR is next step in systematization of solution. I can write this info in RFC.

File: geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java [hunk 1]
```diff
@@ -159,32 +161,32 @@ public int sendTo(SocketChannel chan) throws IOException {
...
+  private void sendChunkTo(ByteBuffer in, SocketChannel sc, ByteBuffer out, NioFilter ioFilter)
+      throws IOException {
     int bytesSent = in.remaining();
     if (in.isDirect()) {
-      flushBuffer(sc, out);
-      while (in.remaining() > 0) {
-        sc.write(in);
-      }
+      flushBuffer(sc, out, ioFilter);
+      writeToSC(sc, in, ioFilter);
```
Comment (agingade) [inline geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java:189]: I am not expert in this area. But from the code changes it looks like we are calling writeToSC twice now (in flushBuffer)...Is this expected?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 2]
```diff
@@ -967,7 +981,8 @@ private void checkForStuckKeys() {
             tmpSel.selectNow(); // clear canceled key
             sc.registerWithSelector2(selector);
           } else {
-            if (tmpsk.isValid() && tmpsk.isReadable()) {
+            if (tmpsk.isValid() && (tmpsk.isReadable() || tmpsk.isWritable())) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:984]: Why did you add isWritable? I think the old code only used non-blocking on reads. Are you also now doing non-blocking writes?
Note this common expression (line 984 and 1129) could be extracted into a method

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 3]
```diff
@@ -1455,22 +1478,24 @@ private void handleNewClientConnection(final Socket socket) throws IOException {
             "Rejected connection from {} because current connection count of {} is greater than or equal to the configured max of {}",
             new Object[] {socket.getInetAddress(), curCnt,
                 maxConnections});
+
         try {
           refuseHandshake(socket.getOutputStream(),
               String.format("exceeded max-connections %s",
                   maxConnections),
-              REPLY_REFUSED);
+              REPLY_REFUSED, ioFilter, socket);
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:1486]: since you are now passing "socket" into refuseHandshake, you no longer need to also pass in socket.getOutputStream()

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationMetadata.java [hunk 4]
```diff
@@ -47,51 +49,67 @@
...
-        dataOutputStream =
-            new VersionedDataOutputStream(unversionedDataOutputStream, clientVersion);
+    InputStream inputStream = null;
+    try {
+      if (ioFilter == null) {
+        inputStream = socket.getInputStream();
       } else {
-        dataInputStream = unversionedDataInputStream;
-        dataOutputStream = unversionedDataOutputStream;
+        inputStream = ioFilter.getInputStream(socket);
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationMetadata.java:68]: shouldn't the initialization of inputStream happen right before the try (instead of after) since all the finally block does is close the inputStream? Also if you then extracted these lines of code into a method you could then have: final InputStream inputStream = getInputStream(socket, ioFilter); try ...

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java [hunk 5]
```diff
@@ -882,6 +1000,12 @@ void readPayloadFields(final int numParts, final int len) throws IOException {
     int readSecurePart = checkAndSetSecurityPart();
 
     int bytesRemaining = len;
+
+    if (this.ioFilter != null) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java:1004]: In some cases we have a filter even if not using SSL, NioPlainEngine. What is the difference between ioFilter being null and it being an instance of NioPlainEngine? Would it be possible to have it never be null? If so, it would be nice to get rid of all these null checks

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java (hunk 2 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:984]: Without this, test are failing, and no connection can not be established.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationMetadata.java (hunk 4 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientRegistrationMetadata.java:68]: Updated

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java (hunk 3 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:1486]: updated

File: geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java [hunk 6]
```diff
@@ -415,5 +417,29 @@ public void write(ByteBufferInputStream.ByteSource source) {
     source.sendTo(this.buffer);
   }
 
+  private void flushBuffer(SocketChannel sc, ByteBuffer out, NioFilter ioFilter)
```
[the remark this answers is not in the document]
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java:420]: deleted old method

File: geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java [hunk 7]
```diff
@@ -415,5 +417,29 @@ public void write(ByteBufferInputStream.ByteSource source) {
...
+      throws IOException {
+    if (out.position() == 0)
+      return;
+    out.flip();
+    writeToSC(sc, out, ioFilter);
+    out.clear();
+  }
+
+  private void writeToSC(SocketChannel sc, ByteBuffer out, NioFilter ioFilter)
```
[the remark this answers is not in the document]
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/HeapDataOutputStream.java:429]: updated

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java [hunk 8]
```diff
@@ -1424,10 +1439,18 @@ private void handleNewClientConnection(final Socket socket) throws IOException {
...
+    final NioFilter ioFilter;
     try {
       if (isSelector()) {
-        communicationMode = getCommunicationModeForSelector(socket);
+        if (socketCreator.forCluster().useSSL()) {
+          ioFilter = createNIOSSLEngine(socket.getChannel());
+          communicationMode = getCommunicationModeForSSLSelector(socket, ioFilter);
+        } else {
+          ioFilter = new NioPlainEngine(bufferPool);
+          communicationMode = getCommunicationModeForSelector(socket);
```
[the remark this answers is not in the document]
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java:1450]: working on it

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java (hunk 5 above)
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java:1004]: I will investigate, but I think there are cases when for socketchannel there are not defined iofilter.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mivanac) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/AcceptorImpl.java]: updated

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java [hunk 9]
```diff
@@ -663,35 +672,53 @@ void flushBuffer() throws IOException {
     cb.clear();
   }
 
+  private void writeToSC(ByteBuffer cb) throws IOException {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java:675]: rename writeToSocketChannel

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java [hunk 10]
```diff
@@ -1161,10 +1288,146 @@ public void receive() throws IOException {
   public void receive(ServerConnection sc, int maxMessageLength, Semaphore dataLimiter,
       Semaphore msgLimiter) throws IOException {
     this.serverConnection = sc;
+    if (sc.getIOFilter() != null) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/Message.java:1291]: I see this pattern in many places. It seems like the null check could be moved into setIOFilter which would simplify all the callers.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java [hunk 11]
```diff
@@ -1307,7 +1312,11 @@ SelectableChannel getSelectableChannel() {
   }
 
   void registerWithSelector2(Selector s) throws ClosedChannelException {
-    getSelectableChannel().register(s, SelectionKey.OP_READ, this);
+    if ((getIOFilter() != null) && (getIOFilter() instanceof NioSslEngine)) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerConnection.java:1315]: I think the "!= null" is not needed since "null instanceof X" always returns false.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java [hunk 12]
```diff
@@ -49,6 +52,7 @@
   private final KnownVersion clientVersion;
 
   private final byte replyCode;
+  private ServerConnection connection;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ServerSideHandshakeImpl.java:55]: make "connection" final

**Review by kirklund — CHANGES_REQUESTED** [review]
Changes look good, but I really want to see some new unit tests and/or integration tests to cover the changes in the classes listed below. I usually focus more on writing unit rather than integration tests. 

Unit tests do not touch filesystem, sockets, or generally do not use real classes other than small data classes outside the class under test. Integration tests can use the filesystem, sockets, or even create DistributedSystem or Cache instances.

The first step of creating a new unit test for an existing class usually involves changing constructors so you can pass in all the dependencies. This can involve adding chained constructors that chain down to a single final constructor that accepts all dependencies which is then annotated with `@VisibleForTesting` so you can pass in mocks for the tests.

Writing tests for existing tests that have no tests can end up being a sizable portion of the work for pull requests like this. I generally just try to add test coverage for the behavior I've actually touched, changed, or added to rather than back-filling tests for the entire class.

Here's the classes that I believe need at least some new unit tests and/or integration test coverage for the changes you've made in the PR:
* AcceptorImpl (AcceptorImplTest and AcceptorImplIntegrationTest)
* CacheClientNotifier (CacheClientNotifierTest)
* ClientRegistrationMetadata (new ClientRegistrationMetadataTest)
* HeapDataOutputStream (HeapDataOutputStreamTest or HeapDataOutputStreamJUnitTest)
* Message (MessageJUnitTest, maybe a new integration test)
* MessageDispatcher (no tests exist)
* Part (no tests exist)
* ServerConnection (ServerConnectionTest and/or ServerConnectionIntegrationTest)
* ServerSideHandshakeImpl (no tests exist)
* SocketMessageWriter (SocketMessageWriterTest or maybe a new SocketMessageWriterIntegrationTest if using real sockets)

**Comment by onichols-pivotal** [comment]
this PR appears to be abandoned, can it be closed?

**Comment by mivanac** [comment]
This PR is on hold.
